### PR TITLE
[SYCL] Move/Refactor device code checks not in check_device_code

### PR DESCRIPTION
**Pull request by ianayl**
This PR:
- Moves device code checks that are not in the `check_device_code` folder into the `check_device_code` folder
- Removed host code from tests that check only for device code using `SYCL_EXTERNAL` functions (as per #14173)

### CODE REVIEW

**Comment by ianayl** [comment]
Note that `sycl/test/extensions/bindless_images.cpp` is also a device code check that should also be moved to `check_device_code`. However, it is not included in this PR as I am still trying to get in contact with people more knowledgeable about the test. This is taking longer than expected, and thus I figured I might as well get this PR through first, before getting to `sycl/test/extensions/bindless_images.cpp` later.

**Comment by ianayl** [comment]
@sergey-semenov Please excuse the ping, this is ready for review. Thanks in advance!

File: sycl/test/check_device_code/no_iostream.cpp
Comment (sergey-semenov) [inline sycl/test/check_device_code/no_iostream.cpp:1]: And this.

File: sycl/test/check_device_code/host_acc_opt.cpp
Comment (sergey-semenov) [inline sycl/test/check_device_code/host_acc_opt.cpp:1]: So does this one.

File: sycl/test/check_device_code/gdb/accessors.cpp
Comment (sergey-semenov) [inline sycl/test/check_device_code/gdb/accessors.cpp:1]: This one checks host code, not device.

**Review by sergey-semenov — COMMENTED** [review]
LGTM overall.

File: sycl/test/check_device_code/gdb/accessors.cpp
Comment (ianayl) [reply sycl/test/check_device_code/gdb/accessors.cpp:1]: I've gone ahead and reverted the changes. Should be good now!

**Review by sergey-semenov — APPROVED** [review]

**Pull request merged**